### ISSUE TITLE
Cannot update a Component while rendering a different component

### ISSUE THREAD

**Issue by timburgess** [urql-graphql/urql#3088]
### Describe the bug

We've used urql successfully for an in-house token & token refresh with auth-exchange. We're trying to add the Azure authentication (provided by [MSAL](https://github.com/AzureAD/microsoft-authentication-library-for-js) but anytime we perform a query with `useQuery` we get the above error in the console:

![Screenshot from 2023-03-15 11-46-37](https://user-images.githubusercontent.com/28915/226506624-6a77c3e0-9b38-4cba-968d-556fc40cad51.png)

We have flagged this with the msal library: see [https://github.com/AzureAD/microsoft-authentication-library-for-js/issues/5796](https://github.com/AzureAD/microsoft-authentication-library-for-js/issues/5796) however they think this may be a urql issue as there is history of similar issues: [https://github.com/urql-graphql/urql/issues/1382](https://github.com/urql-graphql/urql/issues/1382)

### Reproduction

https://github.com/timburgess/simple-query

### Urql version

urql v3.0.3

### Validations

- [X] I can confirm that this is a bug report, and not a feature request, RFC, question, or discussion, for which [GitHub Discussions](https://github.com/urql-graphql/urql/discussions) should be used
- [X] Read the [docs](https://formidable.com/open-source/urql/docs/).
- [X] Follow our [Code of Conduct](https://github.com/urql-graphql/urql/blob/main/CODE_OF_CONDUCT.md)

**Comment by kitten** [comment]
You can read a longer explanation at #3095

We've basically indeed gotten multiple reports, which fall into three buckets; either:
- The issue happens on React Native and is due to some of RN’s promise+render batching (which is a theory)
- The issue happens due to a side-effect, which triggers a query, which triggers an update, all during a render phase of a different component (which is the most common)
- The issue has no reproduction, no details of repercussions, and no further info

Obviously, I'm thankful that you've provided a reproduction, because we don't get these often with this kind of issue. However, I do have to note that the reproduction isn't runnable for me, so if you could update it or provide some instructions that'd be much appreciated (see screenshot attached)

To put a long story short (from the linked PR), it's _always_ safe to ignore the warning. I've never seen reports nor have any proof of it causing any issues.
It does fire in cases where there definitely are issues (like nested `setState(prev => setState(), prev)` calls)

However, it often happens when a side-effect triggers a query, which then triggers an update in a component that is subscribed to this query. It's sometimes extremely hard to follow and reason about why this happens, because these ticks/render cycles are invisible and hard to track.

In your case, it's likely caused by some kind of synchronisation that happens between `msalInstance.acquireTokenSilent` and `MsalProvider`, but it's hard to tell. It's possible it'll just disappear if you either start using a cache or if you add an `await Promise.resolve()` after the aforementioned call 🤷 

Again, this is sometimes hard to reason about. Hence, the PR I've linked (#3095) _silences_ these warnings. It doesn't address the root cause that often causes them because that's out of our control, but it does silence the warning (at the cost of single-render-cycle tearing becoming possible)

It's unlikely that we'd backport the fix to the current release, since we have a major release batch for all packages ready that's about to go out, but if we can get your reproduction to work, I'm happy to go in and test this out for you.

<img width="771" alt="Screenshot 2023-03-24 at 00 32 09" src="https://user-images.githubusercontent.com/2041385/227401555-223a0f27-5e61-4539-9025-75da279469a8.png">

**Comment by timburgess** [comment]
@kitten Thanks Phil! Thats consistent with things working with problems for us but getting repeated console errs. I'll have a poke about on the weekend and see if I can get a better repro.
